presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java

```diff
@@ -59,31 +60,43 @@ public int getPartitionedSplitsOnNode(Node node)
     {
         @GuardedBy("this")
         private final List<RemoteTask> remoteTasks = new ArrayList<>();
+        private final AtomicInteger partitionedSplitCount = new AtomicInteger();
 
-        private synchronized int getPartitionedSplitCount()
+        private int getPartitionedSplitCount()
         {
-            int partitionedSplitCount = 0;
-            for (RemoteTask task : remoteTasks) {
-                partitionedSplitCount += task.getPartitionedSplitCount();
-            }
-            return partitionedSplitCount;
+            return partitionedSplitCount.get();
         }
 
         private synchronized void addTask(RemoteTask task)
         {
             remoteTasks.add(task);
+            updatePartitionedSplitCount(task.getPartitionedSplitCount());
```

Comment:
instead of these calls to `getPartitionedSplitCount()`, I think the `RemoteTask` should guarantee that the listener will observe a sequence of deltas that sums to zero